nomad/testing.go

```diff
@@ -102,19 +104,43 @@ func TestServer(t testing.T, cb func(*Config)) *Server {
 		// Create server
 		server, err := NewServer(config, catalog)
 		if err == nil {
-			return server
+			return server, func() {
+				ch := make(chan error)
+				go func() {
+					defer close(ch)
+
+					// Shutdown server
+					err := server.Shutdown()
+					if err != nil {
+						ch <- errors.Wrap(err, "failed to shutdown server")
+					}
+
+					freeport.Return(ports)
+				}()
+
+				select {
+				case e := <-ch:
+					if e != nil {
+						t.Fatal(e.Error())
+					}
+				case <-time.After(1 * time.Minute):
```

Comment:
Why do this in a goroutine instead of letting go's test timeout detect this? Go's default test timeout is very long, but hopefully blocking for a long time on shutdown is rare?

Reply:
This was boosted from the [client](https://github.com/hashicorp/nomad/commit/06a4b4add2d0a3b265270f7db39ad43af6c08fec) version; sounds like @notnoop got tired of waiting :)

Reply:
i'm +1 on this.  I've done it in client package for two reasons
* the go test timeout is 15 minutes - that's a very long time to wait to learn that Server shutdown stalled and probably CI/we would kill the process before that time elapses.
* Server/client shutdown stalls are corrolated with actual client failures, and waiting for the timeout will have the actual failure masked.  Consider when an AR is stuck and a test condition fails and `t.Fatal` is called; the defer in test triggers client shutdown, which in turn waits for the stuck AR and blocks indefinitely.  Eventually test fails with timeout and without the original t.Fatal output ever emited to stdout!